### PR TITLE
Sort machines deterministically before scaling down

### PR DESCRIPTION
**Pull request by ingvagabund**
**What this PR does / why we need it**:

Sort the machines deterministically to delete machines in the same order. In general the machines are sequenced randomly depending on Machines().List() result. When number of replicas goes down, the reconcilication loop is called multiple times, each time taking a different subset of machines to be deleted. Which may result in deletion of more machines than requested (followed by recreation of new machines). Given each machine represents a node, deleting more nodes than expected can be very disruptive.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note
```

### CODE REVIEW

File: pkg/controller/machineset/controller.go [hunk 1]
```diff
@@ -129,7 +131,22 @@ func (c *MachineSetControllerImpl) Reconcile(machineSet *v1alpha1.MachineSet) er
...
+	}
+
+	// Sort the machines deterministically to delete machines in the same order.
+	// In general the machines are sequenced randomly depending on Machines().List() result.
+	// When number of replicas goes down, the reconcilication loop is called multiple
+	// times, each time taking a different subset of machines to be deleted.
+	// Which may result in deletion of more machines than requested (followed
+	// by recreation of new machines). Given each machine represents a node,
+	// deleting more nodes than expected can be very disruptive.
+	sort.Strings(machineNames)
```
Comment (ingvagabund) [inline pkg/controller/machineset/controller.go]: So far the machines were listed randomly. Given a machine name is generated randomly, the original deletion policy (deleting first machines in a list) is preserved.

**Comment by ingvagabund** [comment]
@frobware @enxebre @roberthbailey @hardikdr

File: pkg/controller/machineset/controller.go (hunk 1 above)
Comment (frobware) [reply pkg/controller/machineset/controller.go]: > Given a machine name is generated randomly, the original deletion policy (deleting first machines in a list) is preserved.

How does the final sort preserve the original deletion policy?

Comment (ingvagabund) [reply pkg/controller/machineset/controller.go]: The original deletion policy takes first items from a list where whose order is generated randomly. Reordering the items based on a name that is generated randomly as well changes order if items from one random sequence to another one.

**Comment by alvaroaleman** [comment]
> 	// When number of replicas goes down, the reconcilication loop is called multiple
	// times, each time taking a different subset of machines to be deleted.
        // Which may result in deletion of more machines than requested (followed
￼ 	// by recreation of new machines

Why is it called multiple times? It gets called once for the change of `.Spec.Replicas` and then once again every time a `machine` gets deleted or am I missing something here?

WTR to the issue that too many machines may be deleted: Wouldn't just filtering out machines in `shouldExcludeMachine` that have a non-nil `deletionTimestamp` fix that?

The only possible issue I see here is a racecondition in scaling down when a `machine` gets deleted by one `Reconsile`, that deletion is not propagated back to the lister yet while another `Reconsile` gets called, resulting in that other `Reconsile` to delete more machines. However this is an issue that can only prevented by using some locking mechanism that blocks further deletions until the first deletions arrived in the lister (And filtering properly, as mentioned above).

Does that make sense or did I get something wrong?

**Comment by roberthbailey** [comment]
Closing as per https://github.com/kubernetes-sigs/cluster-api/pull/558#issuecomment-433380386.

/close

**Pull request closed without merging**